### PR TITLE
Ent Search: avoid generating invalid config in the presence of user of user overrides

### LINKED ISSUE #5290: Enterprise Search controller generates invalid configuration when multiple auth sources are used
When users configure additional authentication sources in ECK for example Active Directory the operator generates an invalid configuration file.

For example following the instructions in https://www.elastic.co/guide/en/enterprise-search/7.16/active-directory-auth.html

Users would set up a configuration like the following in the Enterprise Search manifest:
```
  config:
    ent_search.auth:
      native1:
        source: elasticsearch-native
        order: 1
      my_ad:
        source: elasticsearch-active-directory
        order: 2
```

which leads to a generated configuration as follows: 

```
ent_search:
  auth:
    my_ad:
      order: 2
      source: elasticsearch-active-directory
    native1:
      order: 1
      source: elasticsearch-native
    source: elasticsearch-native
```
which is invalid. 

Omitting the `native` auth source in the Enterprise Search custom resource `config` does not help either. The configuration generated is still invalid.

Affected are versions of the Elastic Stack before 7.17 because of https://github.com/elastic/cloud-on-k8s/pull/5085

### PR DESCRIPTION
**Pull request by pebrc**
Fixes #5290 

The issue as all the details. But in short the changes are: 
* do not generate the `ent_search.auth.source` setting as of ES 7.14
* do not generate the above setting in the presence of user the defined auth settings to avoid generating invalid config (this applies to 7.10-7.13)

I hope I have covered all cases, the risk here is that I missed a combination that would lead to one of the following issues:
1. the auto-migration between 7.13 and 7.14 not working due to the missing setting
2. a variant of a user setting that implicitly requires the generated setting to be present

I think I can exclude both cases based on my research but I still wanted to highlight them here because I find the various possible permutations of settings and Elastic stack versions quite complicated.

Here is my reasoning: 
1. we already defaulted to the `elasticsearch-native` setting before 7.14 so there is nothing to migrate. If a user forcibly went back to `standard` auth this settings override will be preserved regardless
2. pre 7.10 there was only one key `ent_search.auth.source` and `ent_search.auth.name` the latter only valid when SAML was in use, a user had to override `source` anyway to use that. Post 7.10 the new namespaced auth config format was incompatible with the setting we were generating anyway so the fact that we did not have any user reports about this tells me that nobody was using that because if so we would have had this bug report earlier.

### CODE REVIEW

File: pkg/controller/common/settings/canonical_config_test.go [hunk 1]
```diff
@@ -451,3 +451,70 @@ func TestNewCanonicalConfigFrom(t *testing.T) {
...
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.c.HasChildConfig(tt.key); got != tt.want {
+				t.Errorf("HasChildConfig() = %v, want %v", got, tt.want)
+			}
+		})
+	}
```
Comment (barkbay) [inline pkg/controller/common/settings/canonical_config_test.go:519]: If you accept the previous suggestion about checking the error:

```suggestion
		{
			name: "invalid config",
			c: MustCanonicalConfig(map[string]interface{}{
				"x": "{",
			}),
			key:     "x",
			want:    false,
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := tt.c.HasChildConfig(tt.key)
			if (err != nil) != tt.wantErr {
				t.Errorf("HasChildConfig() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("HasChildConfig() = %v, want %v", got, tt.want)
			}
		})
	}
```

**Review by barkbay — APPROVED** [review]
I did a few tests, it seems to work as expected. The reasoning behind this change makes sense to me, I also _think_ the edge cases are covered.

 I just left a minor suggestion on error handling in `HasChildConfig()`.

File: pkg/controller/common/settings/canonical_config.go [hunk 2]
```diff
@@ -153,6 +153,15 @@ func (c *CanonicalConfig) HasKeys(keys []string) []string {
 	return has
 }
 
+// HasChildConfig returns true if c has a child config object below key.
+func (c *CanonicalConfig) HasChildConfig(key string) bool {
+	if c == nil {
+		return false
+	}
+	_, err := c.asUCfg().Child(key, -1, Options...)
+	return err == nil
```
[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/common/settings/canonical_config.go:162]: My thinking was that the two types of error that can be raised here are either ["type mismatch"](https://github.com/elastic/go-ucfg/blob/fc880abbe1f30b653d113da96a4a7e82743c0cc1/error.go#L72)
or ["missing field"](https://github.com/elastic/go-ucfg/blob/fc880abbe1f30b653d113da96a4a7e82743c0cc1/error.go#L62). We want both of those to be interpreted as "this config does not have a child config at this path"

I think the missing path case is self explanatory. The type mismatch would be if in my understand if we have a primitive value a the position the key is pointing to. 

```yaml
a: b
```

or your example
```yaml
x: {
```

I think in these cases I would not expect an error to be raised but just a `false` return value because `b` and `{` are not child configs but primitives. 

Let me know if I misunderstood your example or whether I am missing something (could totally be the case 😄 )

[the remark this answers is not in the document]
Comment (barkbay) [reply pkg/controller/common/settings/canonical_config.go:162]: Ok, got it 👍  You can ignore the suggestion 😄 , maybe just add a small comment to explain why we don't really care about the type of the returned error.

File: pkg/controller/common/settings/canonical_config_test.go (hunk 1 above)
Comment (aymardika) [reply pkg/controller/common/settings/canonical_config_test.go:519]: I am still facing this issue on a standalone docker environment , when I specify this inenterprise-search.yml  (version 8.1.2)
``````
ent_search.auth:
  native1:
    source: elasticsearch-native
    order: 0
``````

Here is the error message 

workspace-kibana-1         | [2022-04-04T03:22:42.080+00:00][FATAL][root] Error: Unknown configuration key(s): "ent_search.auth.native1.source", "ent_search.auth.native1.order". Check for spelling errors and ensure that expected plugins are installed.
workspace-kibana-1         |     at ensureValidConfiguration (/usr/share/kibana/src/core/server/config/ensure_valid_configuration.js:35:11)
workspace-kibana-1         |     at Server.preboot (/usr/share/kibana/src/core/server/server.js:164:5)
workspace-kibana-1         |     at Root.preboot (/usr/share/kibana/src/core/server/root/index.js:48:14)
workspace-kibana-1         |     at bootstrap (/usr/share/kibana/src/core/server/bootstrap.js:99:9)
workspace-kibana-1         |     at Command.<anonymous> (/usr/share/kibana/src/cli/serve/serve.js:216:5)
